temporal-sdk/src/main/java/io/temporal/common/reporter/MicrometerClientStatsReporter.java

```diff
@@ -19,22 +19,44 @@
 
 package io.temporal.common.reporter;
 
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.AtomicDouble;
 import com.uber.m3.tally.Capabilities;
 import com.uber.m3.tally.CapableOf;
 import com.uber.m3.tally.StatsReporter;
 import com.uber.m3.util.Duration;
 import io.micrometer.core.instrument.MeterRegistry;
 import io.micrometer.core.instrument.Tag;
 import io.micrometer.core.instrument.Timer;
-import java.util.Map;
-import java.util.Objects;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.TimeUnit;
 import java.util.stream.Collectors;
 
 public class MicrometerClientStatsReporter implements StatsReporter {
 
   private final MeterRegistry registry;
 
+  /**
+   * This retains all created gauges to keep them publishing. According to Micrometer contract, it's
+   * a user code <a
+   * href="https://micrometer.io/docs/concepts#_why_is_my_gauge_reporting_nan_or_disappearing">
+   * "responsibility to hold a strong reference to the state object that you are measuring with a
+   * Gauge" </a>.
+   *
+   * <p>The need to forcefully retain all the gauges is coming from misalignment of Tally's (that
+   * doesn't expose any way to retain or not retain a gauge) and Micrometer/Prometheus Gauge API.
+   *
+   * <p>This theoretically may cause a memory leak if users create gauges with names or tags of very
+   * large cardinalities or if the names of gauges or tags have unique values.
+   *
+   * <p>The design to retain gauges is based on <a
+   * href="https://github.com/uber-java/tally/blob/master/prometheus/src/main/java/com/uber/m3/tally/experimental/prometheus/PrometheusReporter.java">Tally's
+   * PrometheusReporter<a/>
+   */
+  private final ConcurrentMap<MetricID, AtomicDouble> registeredGauges = new ConcurrentHashMap<>();
```

Comment:
The gauge itself is just an atomic? Odd

Reply:
Well, a Micrometer has a concept of "value measured by a Gauge". You rarely need to work with the Gauge object itself (while you can, using Gauge.builder()). To retain a Gauge, you need to actually retain this "measured value", not the Gauge itself, because by default Gauges have weak references to the "measured value" (which can be changed by passing strongReference=true to the Gauge Builder). Solution with retaining the Gauge itself ends up more complicated and less idiomatic Micrometer.

Reply:
Ahhhh, that makes sense in a roundabout way